File: api/v1/schemas/strategy_validation.py
# -*- coding: utf-8 -*-
"""API contracts for strategy versions and validation scorecards."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict

from src.schemas.strategy_validation import (
    StrategyBenchmarkPolicy,
    StrategyCostModel,
    StrategyTransitionRequest,
    StrategyValidationRun,
    StrategyVersionManifest,
)


class StrategyCreateRequest(StrategyVersionManifest):
    pass


class StrategyRunCreateRequest(StrategyValidationRun):
    pass


class StrategyTransitionBody(StrategyTransitionRequest):
    pass


class StrategyValidationRunItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    run_id: str
    strategy_key: str
    strategy_version: str
    validation_kind: str
    protocol: dict[str, Any]
    dataset_hash: str
    engine_version: str
    status: str
    status_label: str
    qualifying: bool
    result: dict[str, Any]
    run_hash: str
    created_at: str


class StrategyVersionItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    strategy_key: str
    version: str
    name: str
    change_summary: str
    changed_dimension: str
    markets: list[str]
    instrument_types: list[str]
    horizons: list[str]
    evaluation_mode: str
    policy: dict[str, Any]
    cost_model: StrategyCostModel
    benchmark_policy: StrategyBenchmarkPolicy
    status: str
    status_label: str
    allowed_transitions: list[str]
    latest_run: StrategyValidationRunItem | None = None
    manifest_hash: str
    created_at: str


class StrategyVersionListResponse(BaseModel):
    items: list[StrategyVersionItem]


class StrategyTransitionResponse(BaseModel):
    strategy_key: str
    version: str
    from_status: str
    status: str
    status_label: str
    human_reason: str
    transition_id: int

File: apps/dsa-web/src/types/strategyValidation.ts
export type StrategyStatus =
  | 'draft'
  | 'backtest_running'
  | 'backtest_failed'
  | 'simulation'
  | 'small_capital'
  | 'active'
  | 'retired';

export type StrategyValidationStatus = 'completed' | 'failed' | 'unable';

export interface StrategyBucketDimensions {
  horizon?: string;
  market?: string;
  productType?: string;
  instruction?: string;
  marketRegime?: string;
  period?: string;
}

export interface StrategyBucketMetrics {
  sampleCount: number;
  winRatePct?: number | null;
  winDefinition?: string | null;
  netReturnAfterCostPct?: number | null;
  benchmarkExcessPct?: number | null;
  maximumDrawdownPct?: number | null;
  averageGainPct?: number | null;
  averageLossPct?: number | null;
  turnoverPct?: number | null;
  totalCostPct?: number | null;
  unableCount: number;
}

export interface StrategyMetricBucket {
  dimensions: StrategyBucketDimensions;
  metrics: StrategyBucketMetrics;
}

export interface StrategyValidationResult {
  historicalStatus?: string;
  displayMessage?: string;
  eligibleEventCount?: number;
  evaluationCount?: number;
  buckets: StrategyMetricBucket[];
  unableReasons: string[];
  resultHash?: string;
}

export interface StrategyValidationRun {
  runId: string;
  strategyKey: string;
  strategyVersion: string;
  validationKind: 'historical_backtest' | 'forward_observation';
  protocol: Record<string, unknown>;
  datasetHash: string;
  engineVersion: string;
  status: StrategyValidationStatus;
  statusLabel: string;
  qualifying: boolean;
  result: StrategyValidationResult;
  runHash: string;
  createdAt: string;
}

export interface StrategyVersion {
  strategyKey: string;
  version: string;
  name: string;
  changeSummary: string;
  changedDimension: string;
  markets: string[];
  instrumentTypes: string[];
  horizons: string[];
  evaluationMode: 'historical_and_forward' | 'forward_only';
  policy: Record<string, unknown>;
  costModel: Record<string, unknown>;
  benchmarkPolicy: Record<string, unknown>;
  status: StrategyStatus;
  statusLabel: string;
  allowedTransitions: StrategyStatus[];
  latestRun: StrategyValidationRun | null;
  manifestHash: string;
  createdAt: string;
}

export interface StrategyTransitionRequest {
  toStatus: StrategyStatus;
  humanReason: string;
}

export interface StrategyTransitionResponse {
  strategyKey: string;
  version: string;
  fromStatus: StrategyStatus;
  status: StrategyStatus;
  statusLabel: string;
  humanReason: string;
  transitionId: number;
}
